presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -314,18 +315,50 @@ public FileStatus getFileStatus(Path path)
         }
 
         return new FileStatus(
-                getObjectSize(metadata),
+                getObjectSize(metadata, path),
                 false,
                 1,
                 BLOCK_SIZE.toBytes(),
                 lastModifiedTime(metadata),
                 qualifiedPath(path));
     }
 
-    private static long getObjectSize(ObjectMetadata metadata)
+    private long getObjectSize(ObjectMetadata metadata, Path path) throws IOException
     {
         String length = metadata.getUserMetadata().get(UNENCRYPTED_CONTENT_LENGTH);
-        return (length != null) ? Long.parseLong(length) : metadata.getContentLength();
+
+        if (length != null) {
+            return Long.parseLong(length);
+        }
+
+        // check if file has KMS encryption
+        String cryptoAlg = metadata.getUserMetadata().get(CRYPTO_KEYWRAP_ALGORITHM);
+        if (cryptoAlg != null && cryptoAlg.equalsIgnoreCase("kms")) {
+            // KMS client-side encryption
+            long decryptedLength = metadata.getContentLength();
+
+            // load file and determine actual length
+            FSDataInputStream inputStream = this.open(path);
+
+            // read last 25 bytes of file to help determine length
+            int eofOffset = 25;
```

Comment:
I guess the difference of 4-15 bytes come from this:
KMS uses AES [[1]](http://docs.aws.amazon.com/kms/latest/developerguide/crypto-intro.html).  AES, as a block cipher, does not change file size but round the content length to multiple of its block size, which is always 16 bytes [[2]](http://stackoverflow.com/questions/3283787/size-of-data-after-aes-cbc-and-aes-ecb-encryption) [[3]](https://en.wikipedia.org/wiki/Block_cipher). That's why we see difference in 4-15 bytes (though I think we should be seeing difference between 0-3 bytes range as well).

Since KMS claims it uses AES, I guess by checking `x-amz-wrap-alg == "kms"` should be enough? If one day KMS supports more standard that breaks this, we could check `x-amz-cek-alg` to ensure it is AES before we apply the 25 byte rule.